api/src/main/java/com/epam/pipeline/manager/cluster/performancemonitoring/ResourceMonitoringManager.java

```diff
@@ -408,11 +409,21 @@ private void setInstanceTypeMap(final Map<String, InstanceType> instanceTypeMap)
         }
 
         private void processServerlessRuns() {
-            final Integer timeout = preferenceManager.getPreference(SystemPreferences.LAUNCH_SERVERLESS_STOP_TIMEOUT);
-            final LocalDateTime maxLastUpdate = LocalDateTime.now().minusMinutes(timeout);
-            final List<PipelineRun> runs = ListUtils.emptyIfNull(
-                    pipelineRunManager.loadExpiredServerlessRuns(maxLastUpdate));
-            runs.forEach(run -> pipelineRunManager.stopServerlessRun(run.getId()));
+            final List<StopServerlessRun> activeServerlessRuns = ListUtils.emptyIfNull(
+                    pipelineRunManager.loadActiveServerlessRuns());
+            activeServerlessRuns.stream()
+                    .filter(this::serverlessRunIsExpired)
+                    .forEach(run -> pipelineRunManager.stopServerlessRun(run.getId()));
+        }
+
+        private boolean serverlessRunIsExpired(final StopServerlessRun run) {
+            return run.getLastUpdate().isBefore(LocalDateTime.now().minusMinutes(getTimeoutMinutes(run)));
+        }
+
+        private Long getTimeoutMinutes(final StopServerlessRun run) {
+            return Objects.nonNull(run.getStopAfter())
+                    ? run.getStopAfter()
+                    : preferenceManager.getPreference(SystemPreferences.LAUNCH_SERVERLESS_STOP_TIMEOUT).longValue();
```

Comment:
We also need to handle situation when `LAUNCH_SERVERLESS_STOP_TIMEOUT` is null - consider run non-expired in this case

Reply:
Thanks, done